src/fsharp/service/SemanticClassificationKey.fs

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) Microsoft Corporation.  All Rights Reserved.  See License.txt in the project root for license information.
+
+namespace FSharp.Compiler.SourceCodeServices
+
+open System
+open System.IO
+open System.IO.MemoryMappedFiles
+open System.Reflection.Metadata
+
+open FSharp.NativeInterop
+
+open FSharp.Compiler
+open FSharp.Compiler.Range
+
+#nowarn "9"
+
+[<Sealed>]
+type SemanticClassificationKeyStore(mmf: MemoryMappedFile, length) =
+
+    let mutable isDisposed = false
+    let checkDispose() =
+        if isDisposed then
+            raise (ObjectDisposedException("SemanticClassificationKeyStore"))
+
+    member _.ReadRange(reader: byref<BlobReader>) =
+        let startLine = reader.ReadInt32()
+        let startColumn = reader.ReadInt32()
+        let endLine = reader.ReadInt32()
+        let endColumn = reader.ReadInt32()
+        let fileIndex = reader.ReadInt32()
+
+        let posStart = mkPos startLine startColumn
+        let posEnd = mkPos endLine endColumn
+        mkFileIndexRange fileIndex posStart posEnd
+
+    member _.ReadKeyString(reader: byref<BlobReader>) =
+        let size = reader.ReadInt32()
+        let keyString = ReadOnlySpan<char>(reader.CurrentPointer |> NativePtr.toVoidPtr, size)
+        reader.Offset <- reader.Offset + size
+        keyString
+
+    member this.GetAll () =
+        checkDispose()
+        
+        let results = ResizeArray()
+        
+        use view = mmf.CreateViewAccessor(0L, length)
+        let mutable reader = BlobReader(view.SafeMemoryMappedViewHandle.DangerousGetHandle() |> NativePtr.ofNativeInt, int length)
+        
+        reader.Offset <- 0
+        while reader.Offset < reader.Length do
+            let m = this.ReadRange &reader
+            let keyString = this.ReadKeyString &reader
+            let scopt = SemanticClassificationType.FromString (keyString.ToString())
+            match scopt with
+            | None -> ()
+            | Some sc -> results.Add struct (m, sc)
+        
+        results.ToArray()
+
+    interface IDisposable with
+
+            member _.Dispose() =
+                isDisposed <- true
+                mmf.Dispose()
+
+and [<Sealed>] SemanticClassificationKeyStoreBuilder() =
+
+    let b = BlobBuilder()
+
+    let writeString (str: string) =
+        b.WriteUTF16 str
+
+    let writeRange (m: Range.range) =
+        b.WriteInt32(m.StartLine)
+        b.WriteInt32(m.StartColumn)
+        b.WriteInt32(m.EndLine)
+        b.WriteInt32(m.EndColumn)
+        b.WriteInt32(m.FileIndex)
+
+    member _.Write (m: Range.range, sct: SemanticClassificationType) =
+        writeRange m
+
+        let fixup = b.ReserveBytes 4 |> BlobWriter
+
+        let preCount = b.Count
+
+        writeString (sct.ToString())
```

Comment:
Could we make `SemanticClassificationType` an enum and just store an `int32` instead of the string? I know it's a union right now, but I don't see why it couldn't be just an enum.

Assuming that `SemanticClassificationType` becomes an enum and if you think about the data for classification: `struct(range * SemanticClassificationType)` - this is a blittable type and can just be written directly to memory. We have an array of these `struct(range * SemanticClassificationType) []` - that entire array can just be one copy operation to a MMF